doc/architecture/004-Component-architecture.md

```diff
@@ -0,0 +1,48 @@
+# Component architecture
+
+## Context
+
+As previously discussed, NixOps 4's architecture has great similarity with Nix's architecture.
+
+The goal of this document is to outline the components and layers that make up NixOps 4.
+
+## Decision
+
+At a high level, `nixops4` is an executable that realises its extensions through Nix, starts them, and interacts with them.
+
+`nixops4` individually has a layered application architecture, with the following layers:
+
+- The NixOps CLI, analogous to Nix's `src/nix`
+  - Implements the command line interface
+- The NixOps evaluator process
+  - Implements an interface provided by the NixOps core library
+  - Exists as a separate process, to improve the robustness of the system
+    - Data loss following resource creation can have a real world cost, especially if not reported to the user.
+- The NixOps core library, analogous to Nix's `libstore`, and using it, but not `libexpr`.
+  - Implements the interaction with resource implementations
+  - Coordinates the interactions between resources and the evaluator
+
```

Comment:
I'm not left with a clear picture of what the library contains. Based on our conversation:
- bulk of the code that the CLI runs
- code for the evaluator to interact with the CLI